src/main/java/BasicMAPF/Solvers/PIBT/PIBT_Solver.java

```diff
@@ -92,63 +113,58 @@ protected void init(MAPF_Instance instance, RunParameters parameters) {
         initPriority(instance);
 
         // distance between every vertex in the graph to each agent's goal
-        this.heuristic = new DistanceTableAStarHeuristic(instance.agents, instance.map);
+        if (parameters.aStarGAndH instanceof DistanceTableAStarHeuristic) {
+            this.heuristic = (DistanceTableAStarHeuristic) parameters.aStarGAndH;
+        }
+        else {
+            this.heuristic = new DistanceTableAStarHeuristic(instance.agents, instance.map);
+        }
     }
 
 
     @Override
     protected Solution runAlgorithm(MAPF_Instance instance, RunParameters parameters) {
-
         // each iteration of the while represents timestamp
         while (!(finished())) {
 
-            if (checkTimeout()) {
+            // loop detection
+            ArrayList<I_Location> currentConfiguration = new ArrayList<>(instance.agents.size());
+            for (Agent agent : instance.agents) {
+                currentConfiguration.add(this.currentLocations.get(agent));
+            }
+            if (this.configurations.contains(currentConfiguration)) {
+                if (DEBUG >= 2){
+                    System.out.println("LOOP DETECTED");
+                }
```

Comment:
I pushed some changes here. Most importantly, note that you were iterating over a set to create the configuration, meaning the agents were not guaranteed to be the same order every time, so we could have had both false negatives and false positives.